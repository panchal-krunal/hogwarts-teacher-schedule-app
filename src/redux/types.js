export const TOGGLE_TEACHER_PRESENCE = 'TOGGLE_TEACHER_PRESENCE'
export const ALLOCATE_SUBJECT = 'ALLOCATE_SUBJECT'
export const ASSIGN_TEACHER = 'ASSIGN_TEACHER'