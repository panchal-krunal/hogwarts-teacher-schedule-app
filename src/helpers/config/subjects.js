const subjects = {
  gaJoeo1e1K: {
    name: "Potions master",
  },
  ibuUcWMKR5: {
    name: "Defense Against the Dark Arts",
  },
};
export default subjects;