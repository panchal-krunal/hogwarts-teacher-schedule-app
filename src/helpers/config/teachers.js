const teachers = {
  ql_HP3WWb5: {
    name: "Professor Dumbledore",
    subject: "",
    reporting_to: null,
    designation: "Headmaster",
    isPresent: true,
  },
  ihUHzDMgDf: {
    name: "Minerva McGonagall",
    subject: "",
    reporting_to: "ql_HP3WWb5",
    designation: "Headmistress",
    isPresent: true,
  },
  vWljuJ10kr: {
    name: "Rubeus Hagrid",
    subject: "gaJoeo1e1K",
    reporting_to: "ihUHzDMgDf",
    designation: "Standby Professor",
    isPresent: true,
    isHeadingSubject:true
  },
  e6Xah9MirI: {
    name: "Horace Slughorn",
    subject: "gaJoeo1e1K",
    reporting_to: "vWljuJ10kr",
    designation: "",
    isPresent: true,
  },
  SV4mio1jYY: {
    name: "Severus Snape",
    subject: "gaJoeo1e1K",
    reporting_to: "vWljuJ10kr",
    designation: "",
    isPresent: true,
  },
  xhxpdg8RzN: {
    name: "Alastor Moody",
    subject: "ibuUcWMKR5",
    reporting_to: "ihUHzDMgDf",
    designation: "Standby Professor",
    isPresent: true,
    isHeadingSubject:true
  },
  rCkLspfCxE: {
    name: "Remus Lupin",
    subject: "ibuUcWMKR5",
    reporting_to: "xhxpdg8RzN",
    designation: "",
    isPresent: true,
  },
  rCkLspfCxE: {
    name: "Gilderoy Lockhart",
    subject: "ibuUcWMKR5",
    reporting_to: "xhxpdg8RzN",
    designation: "",
    isPresent: true,
  },
};

export default teachers;
