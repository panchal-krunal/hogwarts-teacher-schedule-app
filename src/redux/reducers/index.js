import { combineReducers } from "redux";
import teacherReducer from "./teacherReducer";
import subjectReducer from "./subjectReducer";
import studentReducer from "./studentReducer";
const rootReducer = combineReducers({
  teacherReducer,
  subjectReducer,
  studentReducer
});

export default rootReducer;
