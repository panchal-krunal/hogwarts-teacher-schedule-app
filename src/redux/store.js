import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import reducer from "./reducers";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";

const persistConfig = {
  key: "root",
  storage,
};

const persistedReducer = persistReducer(persistConfig, reducer);

const store = createStore(
  persistedReducer,
  compose(
    applyMiddleware(thunk),
    window.devToolsExtension ? window.devToolsExtension() : (f) => f
  )
);

const persistor = persistStore(store);

export { store, persistor };
