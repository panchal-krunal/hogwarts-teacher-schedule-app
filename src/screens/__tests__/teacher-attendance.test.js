import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import TeacherAttendance from "../teacher-attendance";
import configureStore from "redux-mock-store";
import teachers from "../../helpers/config/teachers";
import subjects from "../../helpers/config/subjects";
import * as ReactReduxHooks from "../../helpers/react-redux-hooks";
// import * as teacherActions from "../../redux/actions/teacherActions";
import thunk from "redux-thunk";
import {TOGGLE_TEACHER_PRESENCE} from '../../redux/types'
Enzyme.configure({ adapter: new Adapter() });

describe("teacher attendance component renders correctly", () => {
  let wrapper;
  let useEffect;
  let store;
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f) => f());
  };
  beforeEach(() => {
    store = mockStore({
      teacherReducer: {
        teachers: { ...teachers },
      },
      subjectReducer: {
        subjects: { ...subjects },
      },
    });
    useEffect = jest.spyOn(React, "useEffect");
    mockUseEffect(); // important to do it twice
    mockUseEffect();
    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    wrapper = mount(<TeacherAttendance />);
  });

  it("should render table", () => {
    const tableElement = wrapper.find("table");
    expect(tableElement).toHaveLength(1);
  });
  it("should render table rows", () => {
    const tableElement = wrapper.find("table");
    const tableRow = tableElement.find("tr");
    expect(tableRow).toHaveLength(8); // 7 teachers and 1 header
  });
  it("should render dropdown for all table rows", () => {
    const tableElement = wrapper.find("table");
    const tableRow = tableElement.find("tr");
    const ddlElement = tableRow.find("select");
    expect(ddlElement).toHaveLength(7); 
  });
  it("should dispatch no action on load", () => {
    const actions = store.getActions();
    expect(actions).toEqual([]);
  });
  it("should dispatch action on dropdown change for present", () => {
    const tableElement = wrapper.find("table");
    const tableRow = tableElement.find("tr");
    const dropdDownElement = tableRow.find("select");
    dropdDownElement.at(0).simulate("change", {
      target: { value: true },
    });
    const expectedAction = [
      {
        type: "TOGGLE_TEACHER_PRESENCE",
        payload: { id: "ql_HP3WWb5", isPresent: true },
      },
    ];
    const actions = store.getActions();
    // console.log(actions);

    expect(actions).toEqual(expectedAction);
  });
  it("should dispatch action on dropdown change for absent", () => {
    const tableElement = wrapper.find("table");
    const tableRow = tableElement.find("tr");
    const dropdDownElement = tableRow.find("select");
    dropdDownElement.at(0).simulate("change", {
      target: { value: false },
    });
    const expectedAction = [
      {
        type: TOGGLE_TEACHER_PRESENCE,
        payload: { id: "ql_HP3WWb5", isPresent: false },
      },
    ];
    const actions = store.getActions();

    expect(actions).toEqual(expectedAction);
  });
});
