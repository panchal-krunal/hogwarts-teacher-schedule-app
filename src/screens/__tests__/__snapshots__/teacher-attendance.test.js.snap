// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teacher attendance component renders correctly should render correctly 1`] = `
<TeacherAttendance>
  <Container
    fluid={false}
  >
    <div
      className="container"
    >
      <ForwardRef
        bordered={true}
        hover={true}
        striped={true}
      >
        <table
          className="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                #
              </th>
              <th>
                Teacher
              </th>
              <th>
                Subject
              </th>
              <th>
                Reporting To
              </th>
              <th>
                Attendance
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              key="ql_HP3WWb5"
            >
              <td>
                1
              </td>
              <td>
                Professor Dumbledore
              </td>
              <td />
              <td />
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="ihUHzDMgDf"
            >
              <td>
                2
              </td>
              <td>
                Minerva McGonagall
              </td>
              <td />
              <td>
                Professor Dumbledore
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="vWljuJ10kr"
            >
              <td>
                3
              </td>
              <td>
                Rubeus Hagrid
              </td>
              <td>
                Potions master
              </td>
              <td>
                Minerva McGonagall
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="e6Xah9MirI"
            >
              <td>
                4
              </td>
              <td>
                Horace Slughorn
              </td>
              <td>
                Potions master
              </td>
              <td>
                Rubeus Hagrid
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="SV4mio1jYY"
            >
              <td>
                5
              </td>
              <td>
                Severus Snape
              </td>
              <td>
                Potions master
              </td>
              <td>
                Rubeus Hagrid
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="xhxpdg8RzN"
            >
              <td>
                6
              </td>
              <td>
                Alastor Moody
              </td>
              <td>
                Defense Against the Dark Arts
              </td>
              <td>
                Minerva McGonagall
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
            <tr
              key="rCkLspfCxE"
            >
              <td>
                7
              </td>
              <td>
                Gilderoy Lockhart
              </td>
              <td>
                Defense Against the Dark Arts
              </td>
              <td>
                Alastor Moody
              </td>
              <td>
                <select
                  onChange={[Function]}
                  value={true}
                >
                  <option
                    value={false}
                  >
                    Absent
                  </option>
                  <option
                    value={true}
                  >
                    Present
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </ForwardRef>
    </div>
  </Container>
</TeacherAttendance>
`;
